src/vm/amd64/CrtHelpers.asm

```diff
@@ -13,516 +13,264 @@
-        jnz     short mset50            ; if nz, more bytes to fill
-mset60: 
-        ; for some reason the assembler doesn't like the REPRET macro on the same line as a label
-        REPRET                          ; return
-
-;
-; Fill 64-byte blocks.
-;
-
-        align   16
-
-        db      066h, 066h, 066h, 090h
-        db      066h, 066h, 090h
-
-mset70: cmp     r9, CACHE_LIMIT_MEMSET / 64    ; check if large fill
-        jae     short mset90            ; if ae, large fill
-mset80: mov     [rcx], rdx              ; fill 64-byte block
-        mov     8[rcx], rdx             ;
-        mov     16[rcx], rdx            ;
-        add     rcx, 64                 ; advance to next block
-        mov     (24 - 64)[rcx], rdx     ;
-        mov     (32 - 64)[rcx], rdx     ;
-        dec     r9                      ; decrement loop count
-        mov     (40 - 64)[rcx], rdx     ;
-        mov     (48 - 64)[rcx], rdx     ;
-        mov     (56 - 64)[rcx], rdx     ;
-        jnz     short mset80            ; if nz, more 64-byte blocks
-        jmp     short mset20            ; finish in common code
-
-;
-; Fill 64-byte blocks nontemporal.
-;
-
-        align
-
-mset90: movnti  [rcx], rdx              ; fill 64-byte block
-        movnti  8[rcx], rdx             ;
-        movnti  16[rcx], rdx            ;
-        add     rcx, 64                 ; advance to next block
-        movnti  (24 - 64)[rcx], rdx     ;
-        movnti  (32 - 64)[rcx], rdx     ;
-        dec     r9                      ; decrement loop count
-        movnti  (40 - 64)[rcx], rdx     ;
-        movnti  (48 - 64)[rcx], rdx     ;
-        movnti  (56 - 64)[rcx], rdx     ;
-        jnz     short mset90            ; if nz, move 64-byte blocks
-   lock or      byte ptr [rsp], 0       ; flush data to memory
-        jmp     mset20                  ; finish in common code
+        mov     r9, 0101010101010101h   
+        imul    rdx, r9                 ; rdx is 8 bytes filler
+
+        cmp     r8, 16                  
+        jbe     mset04                 
+
+        cmp     r8, 512                 
+        jbe     mset00 
+        
+        ; count > 512
+        mov     r10, rcx                ; save dst address
+        mov     r11, rdi                ; save rdi
+        mov     eax, edx                ; eax is value
+        mov     rdi, rcx                ; rdi is dst
+        mov     rcx, r8                 ; rcx is count
+        rep     stosb
+        mov     rdi, r11                ; restore rdi
+        mov     rax, r10
+        ret
+
+        align 16
+mset00: mov     rax, rcx                ; save dst address
+        movd    xmm0, rdx				
+        punpcklbw xmm0, xmm0            ; xmm0 is 16 bytes filler
+
+        cmp     r8, 128                
+        jbe     mset02  
+
+        ; count > 128 && count <= 512
+        mov     r9, r8
+        shr     r9, 7                   ; count/128
+        
+        align 16
+mset01: movdqu	[rcx], xmm0
```

Comment:
Thanks for pointing it out. It is safe to use movdqu because movdqa can raise exception if the address is not aligned. Also, we cannot see much difference between movdqa and movdqu when running experiments.